### ISSUE TITLE
Use Kakoune's history expansion to implement incremental buffer synchronization

### ISSUE THREAD

**Issue by tadeokondrak** [kak-lsp/kak-lsp#552]
For semantic tokens to work on large files on every keystroke, there's no way to make it fast without sending updates incrementally. I think we can use Kakoune's `%val{history}` and `%val{history_id}` for quick synchronization of buffers between Kakoune and kak-lsp.

I've wanted to do it for a long time, but it seems *hard*.
I'm opening this issue in hope that someone else is also interested or has comments/feedback about this.

**Comment by krobelus** [comment]
The naive implementation of semantic-tokens with NormalIdle hooks indeed sends potentially sends a request on every key stroke, even if we didn't modify the buffer.
However, we already have a mechanism in place to only send something whenever the buffer actually changed - this is how we send `textDocument/didChange`. It won't hurt to add a conditional shell block there to also update semantic tokens at the same time.

In case that's not enough, we might need to look further.
`%val{history}` sounds very nice, I didn't know it existed. So the idea is to only send edits/diffs to kak-lsp; then kak-lsp can compute the buffer contents. Hard to tell what the bottlenecks are, I'll try to find out eventually.

**Comment by tadeokondrak** [comment]
> However, we already have a mechanism in place to only send something whenever the buffer actually changed - this is how we send textDocument/didChange. It won't hurt to add a conditional shell block there to also update semantic tokens at the same time.

Right, I have actually tested this out, and there is still a problem on a ~2000 line C file with semantic highlighting with clangd for every keystroke typed in insert mode. This is the delay I'm looking to improve, and it still involves the buffer changing.

The problem might not be here, though. It could be from Kakoune processing the *response* to the semantic tokens request, in which case I can think of two ways to help:

- Maybe there is an issue with the performance of sending any command to Kakoune every keystroke, and https://github.com/kak-lsp/kak-lsp/issues/553 could help.
- Maybe the bottleneck is really how big the semantic tokens update command is, and it could be improved with requesting only part of it. Now that I think about it, incrementally updating semantic tokens probably doesn't actually require doing incremental buffer updates.

**Comment by sclu1034** [comment]
> So the idea is to only send edits/diffs to kak-lsp; then kak-lsp can compute the buffer contents
Depending on the server implementation, things could stop after the first part. The LSP spec includes facilities for requesting tokens only on specific file ranges or diffs/deltas.
